README.md
# rusty-diceware - a password generator using wordlists

Commandline [Diceware][diceware] ([Wayback Machine mirror][diceware-wayback]), with or without dice, written in [rustlang][rustlang].

Please use [Gitlab][gitlab-mirror] for anything whatsoever. Github is just a mirror.

[CHANGELOG here](/CHANGELOG.md).

Inspired by the great passphrase generating solution [Diceware][diceware] ([Wayback Machine mirror][diceware-wayback]) invented by [Arnold G. Reinhold][arnold] ([Wayback Machine mirror][arnold-wayback]) and by Randall Monroe’s [xkcd#936][xkcd-936]:

![“Hidden” alt text jokes are a pain in the ass.](/bin/imgs.xkcd.com/comics/password_strength.png)

## Usage:

```
Usage: diceware [options]

Options:
    -h, --help          This help message.
    -e, --entropy       Display number of entropy bits.
    -r, --dicerolls     Provide results of physical dice rolls. Word per line,
                        same digit order as in the files, digits between and
                        including 1 and 6.
    -n, --nword NWORD   Number of words in a passphrase.
    -d, --delimiter DELIM
                        The delimiter character used to separate the words.
    -f, --wordlist-file FILE
                        Path to a wordlist file.
    -l, --wordlist WORDLIST
                        Wordlist to use. (efflong (default), effshort1,
                        effshort2, minilock, reinhold, or beale)
```

## Featuring

* The three wordlists mentioned in EFF's [Diceware Guide][eff-diceware-guide]:
    * [EFF Long Wordlist][eff-long-wordlist].
    * [EFF Short Wordlist #1][eff-short-wordlist-1].
    * [EFF Short Wordlist #2][eff-short-wordlist-2-0].
* The original [Reinhold wordlist][reinhold-wordlist-asc] ([Wayback Machine mirror][reinhold-wordlist-asc-wayback]).
* The [Beale wordlist][beale-wordlist-asc] ([Wayback Machine mirror][beale-wordlist-asc-wayback]).
* The [MiniLock][minilock] ([github][minilock-github])wordlist. (found in the [phrase.js][minilock-phrase-js] file)
* The all new `--wordlist-file` command line option which loads and uses your very own newline delimited wordlist file. Inquire within!
* Physical dice roll! You can (don't use echo, it will show up in `ps` and show your rolls to other users):

    ```
    $ cat | diceware -l efflong -r
    111111
    111112
    ^D
    abacus abdomen
    ```
* Wordlists live in their own [diceware_wordlists][diceware_wordlists] crate you can use for your own wicked deeds.

## Mirrors

* [Gitlab][gitlab-mirror] is the main repository.
* [Github][github-mirror] is just a mirror.

## Say hello, chat, and/or lurch, aka "community"

You can say hello and/or rant about how terrible rusty-diceware is in the [#rusty-diceware:matrix.org](https://matrix.to/#/#rusty-diceware:matrix.org) chat room.

## Tips

If you want to tip for this work, I have set up a Liberapay account:

<script src="https://liberapay.com/yuvallanger/widgets/button.js"></script>
<noscript><a href="https://liberapay.com/yuvallanger/donate"><img alt="Donate using Liberapay" src="https://liberapay.com/assets/widgets/donate.svg"></a></noscript>

[gitlab-mirror]: <https://gitlab.com/yuvallanger/rusty-diceware/>
[github-mirror]: <https://github.com/yuvallanger/rusty-diceware/>

[diceware_wordlists]: <https://crates.io/crates/diceware_wordlists>

[eff-diceware-guide]: <https://www.eff.org/dice>
[eff-long-wordlist]: <https://www.eff.org/files/2016/07/18/eff_large_wordlist.txt>
[eff-short-wordlist-1]: <https://www.eff.org/files/2016/09/08/eff_short_wordlist_1.txt>
[eff-short-wordlist-2-0]: <https://www.eff.org/files/2016/09/08/eff_short_wordlist_2_0.txt>

[arnold]: <https://theworld.com/~reinhold/>
[diceware]: <https://theworld.com/~reinhold/diceware.html>

[arnold-wayback]: <https://web.archive.org/web/20220608141106/https://theworld.com/~reinhold/>
[diceware-wayback]: <https://web.archive.org/web/20220817092807/https://theworld.com/~reinhold/diceware.html>

[beale-wordlist-asc]: <https://theworld.com/~reinhold/beale.wordlist.asc>
[reinhold-wordlist-asc]: <https://theworld.com/~reinhold/diceware.wordlist.asc>
[minilock-phrase-js]: <https://github.com/kaepora/miniLock/blob/71dcf431886068c9ec7f563c3e4158153229b202/src/js/lib/phrase.js>

[beale-wordlist-asc-wayback]: <https://web.archive.org/web/20220602072646/https://theworld.com/~reinhold/beale.wordlist.asc>
[reinhold-wordlist-asc-wayback]: <https://web.archive.org/web/20220820102521/https://theworld.com/~reinhold/diceware.wordlist.asc>


[rustlang]: <http://rust-lang.org>

[xkcd-936]: <https://www.explainxkcd.com/wiki/index.php/936>

[minilock]: <http://minilock.io>
[minilock-github]: <https://github.com/kaepora/miniLock/>


CHANGELOG.md
# v0.5.8

* Move tests into tests directory.
* Use a reproducible `rand_chacha::ChaCha12Rng` instead of an `rand::prelude::StdRng`. In the current `rand` version, `0.8.5`, `StdRng` is `ChaCha12Rng`, but it may change in the future, as written in <https://docs.rs/rand/0.8.5/rand/rngs/struct.StdRng.html>.

# v0.5.7

Bumping `diceware_wordlists` version in the dependencies.

## v0.5.4 to v0.5.6

* Move wordlists to their own permissively licensed `diceware_wordlists` crate.

## v0.5.3

* Check for legal die digit (not less than 1 not more than 6).

## v0.5.2

* Move code from `/src/bin/diceware.rs`, `/tests/tests.rs`, and `/src/lib.rs` into `/src/main.rs`.

## v0.5.1

* Add physical dice roll input flag (`-r` / `--dicerolls`). Accepts input from standard input.

## v0.5.0

* Add the EFF wordlists as shown in <https://www.eff.org/dice>.
* Make EFF Long Wordlist the default.
* Remove the `--beale` / `--reinhold` / `--minilock` flags.

## v0.4.0

* Add the `-l` / `--wordlist` option.
* Warn that the `--beale` / `--reinhold` / `--minilock` flags would be deprecated in the next minor version.

## v0.3.9

* Move some code from main function to its own function in `src/lib.rs`.
* Remove vestigial naming used when we were stupidly overusing the type system.
* Add test for the MiniLock wordlist.
* Move predefined wordlists to their own `diceware::wordlists` module.
* Add a note about the `--wordlist-file` commandline option.

## v0.3.8

* Switch from a the old and very very weird Struct definition per wordlist to the braindead simple array of `&str`s.
* Add a command line option that takes a filepath to a newline delimited wordlist file so that you could use your own wordlist.

## v0.3.7

* Deduplicate code implementing the `Word` trait using macros. (Thanks @vbrandl!)
* Deduplicate tests code using macros. (Thanks @vbrandl, again!)
* Deduplicate `build.rs` code using a function. How not exciting.

## v0.3.6

* No, really, fix tests. (Added the `Word` trait)

## v0.3.5

* Fix tests. (Was broken due to a changed ThreadRng output)

## v0.3.4

* Deduplicate word printing code by using the trait `Word` and moving that code into a a generic function which uses that trait.
* Move to 2021 edition of Rustlang.  Add a commented out wordlist command line option for a future version.
* Remove some commented out code.
* Use the `Self` keyword in `impl`s instead of the type's concrete name.
* Add this CHANGELOG.


src/tests.rs
extern crate rand;

use rand::{Rng, SeedableRng, StdRng};

use crate::diceware::{BealeWord, ReinholdWord};

fn make_beale_vector() -> Vec<BealeWord> {
    let seed: [u8; 32] = [0; 32];
    let mut rng: StdRng = SeedableRng::from_seed(seed);

    let mut beale_vector: Vec<BealeWord> = vec![];
    for _ in 0..4 {
        let word: BealeWord = rng.gen();
        beale_vector.push(word);
    }
    beale_vector
}

fn make_reinhold_vector() -> Vec<ReinholdWord> {
    let seed: [u8; 32] = [0; 32];
    let mut rng: StdRng = SeedableRng::from_seed(seed);

    let mut reinhold_vector: Vec<ReinholdWord> = vec![];
    for _ in 0..4 {
        let word: ReinholdWord = rng.gen();
        reinhold_vector.push(word);
    }
    reinhold_vector
}

#[test]
fn beale_rng_test() {
    let wanted: Vec<BealeWord> = vec!["dr", "raced", "pvc", "moon"]
        .into_iter()
        .map(BealeWord::new)
        .collect();

    let got = make_beale_vector();

    assert_eq!(got, wanted);
}

#[test]
fn reinhold_rng_test() {
    let wanted: Vec<ReinholdWord> = vec!["douse", "qo", "prune", "moan"]
        .into_iter()
        .map(ReinholdWord::new)
        .collect();

    let got = make_reinhold_vector();

    assert_eq!(got, wanted);
}


tests/tests.rs
use rand_chacha;

macro_rules! create_test {
    ( $wordlist_name: ident, $test_name: ident, $expected: expr ) => {
        #[test]
        fn $test_name() {
            use diceware_wordlists::$wordlist_name;
            use rand::prelude::SeedableRng;
            use rand::prelude::SliceRandom;
            use rand_chacha::ChaCha12Rng;

            fn make_vektor<'a>() -> Vec<&'a str> {
                let seed: [u8; 32] = [0; 32];
                let mut rng: ChaCha12Rng = SeedableRng::from_seed(seed);

                let mut vector: Vec<&str> = vec![];
                for _ in 0..4 {
                    let word: &str = $wordlist_name.choose(&mut rng).unwrap();
                    vector.push(word);
                }
                vector
            }
            let wanted: Vec<&str> = $expected.into_iter().collect();

            let got = make_vektor();

            assert_eq!(got, wanted);
        }
    };
}

create_test!(
    BEALE_WORDLIST,
    beale_rng_test,
    vec!["io", "gavel", "beam", "time"]
);

create_test!(
    REINHOLD_WORDLIST,
    reinhold_rng_test,
    vec!["india", "gamma", "bcd", "theme"]
);

create_test!(
    MINILOCK_WORDLIST,
    minilock_rng_test,
    vec!["hoed", "femininity", "bedsit", "stabbings"]
);

create_test!(
    EFF_LONG_WORDLIST,
    eff_long_rng_test,
    vec!["helpless", "fever", "blooming", "sublease"]
);
create_test!(
    EFF_SHORT_WORDLIST_1,
    eff_short_1_rng_test,
    vec!["five", "boat", "shape", "bony"]
);
create_test!(
    EFF_SHORT_WORDLIST_2_0,
    eff_short_2_0_rng_test,
    vec!["family", "aseptic", "renovator", "atlas"]
);


diceware_wordlists/tests/tests.rs
macro_rules! create_test {
    (
        $wordlist_name: ident,
        $test_name: ident,
        $expected_indices: expr,
        $expected_values: expr,
    ) => {
        #[test]
        fn $test_name() {
            use diceware_wordlists::$wordlist_name;

            let wanted: Vec<&str> = $expected_values.into_iter().collect();

            let got: Vec<&str> = $expected_indices
                .into_iter()
                .map(|i| $wordlist_name[i])
                .collect();

            assert_eq!(got, wanted);
        }
    };
}

#[cfg(feature = "beale")]
create_test!(
    BEALE_WORDLIST,
    beale_test,
    vec![
        0,
        BEALE_WORDLIST.len() - 1,
        490,
        811,
        970,
        1187,
        1726,
        2158,
        2492,
        3005,
        3713,
        4143,
        4289,
    ],
    vec![
        "a", "@", "balls", "boor", "bunts", "cherub", "dime", "ezra", "fuzzy", "hire", "leeway",
        "mills", "ms",
    ],
);

#[cfg(feature = "reinhold")]
create_test!(
    REINHOLD_WORDLIST,
    reinhold_test,
    vec![
        0,
        REINHOLD_WORDLIST.len() - 1,
        490,
        811,
        970,
        1187,
        1726,
        2158,
        2492,
        3005,
        3713,
        4143,
        4289,
    ],
    vec![
        "a", "@", "balk", "bony", "bum", "charm", "dice", "excess", "fuchs", "hg", "le", "mercy",
        "morsel",
    ],
);

#[cfg(feature = "minilock")]
create_test!(
    MINILOCK_WORDLIST,
    minilock_test,
    vec![
        0,
        MINILOCK_WORDLIST.len() - 1,
        7625,
        9287,
        18137,
        26351,
        27726,
        32096,
        37667,
        38151,
        46685,
        47351,
        52997,
    ],
    vec![
        "aardvark",
        "zulus",
        "censer",
        "colonels",
        "exonerate",
        "infringe",
        "jaunts",
        "miserliness",
        "platitudes",
        "pontoon",
        "skua",
        "snuggles",
        "tributes",
    ],
);

#[cfg(feature = "efflong")]
create_test!(
    EFF_LONG_WORDLIST,
    eff_long_test,
    vec![
        0,
        EFF_LONG_WORDLIST.len() - 1,
        490,
        811,
        970,
        1187,
        1726,
        2158,
        2492,
        3005,
        3713,
        4143,
        4289,
    ],
    vec![
        "abacus",
        "zoom",
        "barbecue",
        "candle",
        "charbroil",
        "commodity",
        "deserving",
        "eleven",
        "fame",
        "grew",
        "limpness",
        "nifty",
        "outer",
    ],
);
#[cfg(feature = "effshort1")]
create_test!(
    EFF_SHORT_WORDLIST_1,
    eff_short_1_test,
    vec![
        0,
        EFF_SHORT_WORDLIST_1.len() - 1,
        201,
        250,
        371,
        400,
        565,
        606,
        826,
        890,
        909,
        922,
        966,
    ],
    vec![
        "acid", "zoom", "clerk", "crook", "elbow", "faced", "hut", "kilt", "putt", "rover",
        "sandy", "scary", "shove",
    ],
);
#[cfg(feature = "effshort2")]
create_test!(
    EFF_SHORT_WORDLIST_2_0,
    eff_short_2_0_test,
    vec![
        0,
        EFF_SHORT_WORDLIST_2_0.len() - 1,
        201,
        250,
        371,
        400,
        565,
        606,
        826,
        890,
        909,
        922,
        966,
    ],
    vec![
        "aardvark",
        "zucchini",
        "circle",
        "cytoplasm",
        "enforcer",
        "etiquette",
        "hybrid",
        "jamboree",
        "omnivorous",
        "plywood",
        "pry",
        "puzzle",
        "riches",
    ],
);


src/main.rs
/*
rusty-diceware - a password / passphrasse generator using wordlists, with or without dice.
Copyright (C) 2015-2022  Yuval Langer

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

extern crate getopts;
extern crate rand;

use std::fs::File;
use std::io::BufRead;
use std::io::Read;
use std::process::exit;
use std::str::FromStr;

use diceware_wordlists::Wordlist;
use getopts::Options;
use rand::prelude::SliceRandom;
use rand::rngs::ThreadRng;
use rand::thread_rng;

fn make_options() -> Options {
    let mut opts = Options::new();
    opts.optflag("h", "help", "This help message.");
    opts.optflag("e", "entropy", "Display number of entropy bits.");
    opts.optflag("r", "dicerolls", "Provide results of physical dice rolls. Word per line, same digit order as in the files, digits between and including 1 and 6.");
    opts.optopt("n", "nword", "Number of words in a passphrase.", "NWORD");
    opts.optopt(
        "d",
        "delimiter",
        "The delimiter character used to separate the words.",
        "DELIM",
    );
    opts.optopt("f", "wordlist-file", "Path to a wordlist file.", "FILE");
    opts.optopt(
        "l",
        "wordlist",
        "Wordlist to use. (efflong (default), effshort1, effshort2, minilock, reinhold, or beale)",
        "WORDLIST",
    );
    opts
}

fn print_usage(program: &str, opts: Options) {
    let brief = format!("Usage: {} [options]", program);
    print!("{}", opts.usage(&brief));
}

fn unknown_wordlist(wordlist_name: &str) -> ! {
    eprintln!(
        "Unknown wordlist: {}. Available wordlists: efflong (default), effshort1, effshort2, beale, reinhold, or minilock.",
        wordlist_name,
    );
    exit(-1)
}

fn find_number_of_rolls_needed(wordlist_length: u32) -> u32 {
    let mut x = wordlist_length as u64;
    let mut nrolls: u32 = 0;
    while x != 0 {
        x /= 6;
        nrolls += 1;
    }
    nrolls
}

fn rolls_to_word_index(number_of_rolls_needed: u32, rolls: &[u8]) -> usize {
    if number_of_rolls_needed == rolls.len() as u32 {
        let mut word_number = 0;
        for (i, roll) in rolls.iter().rev().enumerate() {
            if *roll < 1 || *roll > 6 {
                panic!(
                    "Must be a die roll result between and including 1 and 6, not {}",
                    roll
                );
            }
            word_number += ((roll - 1) as usize) * 6_usize.pow(i as u32);
        }
        word_number
    } else {
        panic!(
            "Wrong number of die casts: {}. Needs to be {} die casts.",
            rolls.len(),
            number_of_rolls_needed
        )
    }
}

fn print_words_rolls(
    wordlist: &[&str],
    delimiter: &char,
    is_entropy_printed: &bool,
    rolls: &[&[u8]],
) {
    let number_of_rolls_needed = find_number_of_rolls_needed(wordlist.len() as u32);
    for roll_index in 0..(rolls.len() - 1) {
        let roll_sum = rolls_to_word_index(
            number_of_rolls_needed,
            rolls
                .get(roll_index)
                .unwrap_or_else(|| panic!("Bad roll index: {}", roll_index)),
        );
        let word = wordlist
            .get(roll_sum)
            .unwrap_or_else(|| panic!("Wrong word index: {}", roll_sum));
        print!("{}{}", &word, delimiter);
    }
    let roll_sum = rolls_to_word_index(number_of_rolls_needed, rolls.last().unwrap());
    let word = wordlist.get(roll_sum).unwrap();
    print!("{}", &word);

    println!();
    if *is_entropy_printed {
        println!("{}", entropyn(wordlist, rolls.len() as u64))
    }
}

fn read_rolls() -> Vec<Vec<u8>> {
    let stdin = std::io::stdin();
    let mut rolls: Vec<Vec<u8>> = Vec::new();
    let mut last_number_of_dice = None;
    for line in stdin.lock().lines() {
        let line_value = line.unwrap();
        let line_value_trimmed = line_value.trim();
        if line_value_trimmed.is_empty() {
            continue;
        }
        let current_number_of_dice = line_value_trimmed.len();
        if let Some(last_number_of_dice_value) = last_number_of_dice {
            if last_number_of_dice_value != current_number_of_dice {
                panic!("Not all dice rolls were of the same number of die.");
            } else {
                last_number_of_dice = Some(current_number_of_dice);
            }
        }

        rolls.push(
            line_value_trimmed
                .chars()
                .map(|c| {
                    c.to_string()
                        .parse()
                        .unwrap_or_else(|e| panic!("Not a digit: \"{}\". Error: {}", c, e))
                })
                .collect(),
        );
    }
    rolls
}

fn entropy(wordlist: &[&str]) -> f64 {
    (wordlist.len() as f64).log2()
}

fn entropyn(wordlist: &[&str], n: u64) -> f64 {
    entropy(wordlist) * (n as f64)
}

fn print_words_rng(
    wordlist: &[&str],
    word_num: &u64,
    delimiter: &char,
    is_entropy_printed: &bool,
    rng: &mut ThreadRng,
) {
    for _ in 0..(word_num - 1) {
        let word = wordlist.choose(rng).unwrap();
        print!("{}{}", &word, delimiter);
    }
    let word = wordlist.choose(rng).unwrap();
    print!("{}", word);

    println!();
    if *is_entropy_printed {
        println!("{}", entropyn(wordlist, *word_num))
    }
}

fn load_wordlist_file(filepath: &str) -> String {
    let mut wordlist_file = match File::open(&filepath) {
        Ok(ok) => ok,
        Err(err) => panic!("Unable to open file: {}; due to error: {}", filepath, err),
    };
    let mut wordlist_string = String::new();
    if let Err(err) = wordlist_file.read_to_string(&mut wordlist_string) {
        panic!("Unable to read file: {}; due to error: {}", filepath, err)
    }
    wordlist_string
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let program = &args[0];

    let opts = make_options();

    let matches = match opts.parse(&args[1..]) {
        Ok(m) => m,
        Err(f) => {
            println!("{}\n", f);
            print_usage(program, opts);
            exit(-1);
        }
    };

    if matches.opt_present("h") {
        print_usage(program, opts);
        return;
    };

    let is_physical_rolls = matches.opt_present("r");

    let word_num: u64 = matches
        .opt_str("n")
        .map_or(8, |n_str| n_str.parse::<u64>().ok().unwrap());

    let delimiter: char = matches
        .opt_str("d")
        .map_or(' ', |n_str| n_str.parse::<char>().ok().unwrap());

    let is_entropy_printed = matches.opt_present("entropy");

    let wordlist_name = if let Some(wordlist_option) = matches.opt_str("l") {
        match Wordlist::from_str(&wordlist_option.to_lowercase()) {
            Ok(list) => list,
            _ => unknown_wordlist(&wordlist_option),
        }
    } else {
        Wordlist::default()
    };

    let mut rng = thread_rng();

    if let Some(wordlist_filepath) = matches.opt_str("f") {
        let wordlist_string = load_wordlist_file(&wordlist_filepath);
        let wordlist = wordlist_string
            .split('\n')
            .map(|x| x.trim())
            .filter(|x| x != &"")
            .collect::<Vec<&str>>();

        if is_physical_rolls {
            let rolls = read_rolls();

            print_words_rolls(
                &wordlist,
                &delimiter,
                &is_entropy_printed,
                &rolls.iter().map(|x| x.as_ref()).collect::<Vec<&[u8]>>(),
            );
        } else if word_num != 0 {
            print_words_rng(
                &wordlist,
                &word_num,
                &delimiter,
                &is_entropy_printed,
                &mut rng,
            );
        };
    } else {
        let wordlist = wordlist_name.get_list();

        if is_physical_rolls {
            let rolls = read_rolls();

            print_words_rolls(
                wordlist,
                &delimiter,
                &is_entropy_printed,
                &rolls.iter().map(|x| x.as_ref()).collect::<Vec<&[u8]>>(),
            );
        } else if word_num != 0 {
            print_words_rng(
                wordlist,
                &word_num,
                &delimiter,
                &is_entropy_printed,
                &mut rng,
            );
        };
    };
}


src/diceware.rs
extern crate rand;

use rand::seq::SliceRandom;
use std::fmt;

include!(concat!(env!("OUT_DIR"), "/diceware.rs"));

#[derive(Debug, Clone, Eq, PartialEq, Copy)]
pub struct BealeWord(&'static str);

#[derive(Debug, Clone, Eq, PartialEq, Copy)]
pub struct ReinholdWord(&'static str);

#[derive(Debug, Clone, Eq, PartialEq, Copy)]
pub struct MiniLockWord(&'static str);

impl BealeWord {
    #[cfg(test)]
    pub fn new(word: &'static str) -> BealeWord {
        BealeWord(word.clone())
    }

    pub fn entropy() -> f64 {
        (BEALE_WORDLIST.len() as f64).log2()
    }

    pub fn entropyn(n: u64) -> f64 {
        BealeWord::entropy() * (n as f64)
    }
}

impl ReinholdWord {
    #[cfg(test)]
    pub fn new(word: &'static str) -> ReinholdWord {
        ReinholdWord(word.clone())
    }

    pub fn entropy() -> f64 {
        (REINHOLD_WORDLIST.len() as f64).log2()
    }

    pub fn entropyn(n: u64) -> f64 {
        ReinholdWord::entropy() * (n as f64)
    }
}

impl MiniLockWord {
    #[cfg(test)]
    pub fn new(word: &'static str) -> MiniLockWord {
        MiniLockWord(word.clone())
    }

    pub fn entropy() -> f64 {
        (MINILOCK_WORDLIST.len() as f64).log2()
    }

    pub fn entropyn(n: u64) -> f64 {
        MiniLockWord::entropy() * (n as f64)
    }
}

impl rand::distributions::Distribution<BealeWord> for rand::distributions::Standard {
    fn sample<R: rand::Rng + ?Sized>(&self, mut rng: &mut R) -> BealeWord {
        *BEALE_WORDLIST.choose(&mut rng).unwrap()
    }
}

/*
impl rand::Rand for BealeWord {
    fn rand<R: rand::Rng>(rng: &mut R) -> BealeWord {
        rng.choose(&BEALE_WORDLIST).unwrap().clone()
    }
}
*/

impl fmt::Display for BealeWord {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let BealeWord(w) = self;
        write!(f, "{}", w)
    }
}

impl rand::distributions::Distribution<ReinholdWord> for rand::distributions::Standard {
    fn sample<R: rand::Rng + ?Sized>(&self, mut rng: &mut R) -> ReinholdWord {
        *REINHOLD_WORDLIST.choose(&mut rng).unwrap()
    }
}

/*
impl rand::Rand for ReinholdWord {
    fn rand<R: rand::Rng>(rng: &mut R) -> ReinholdWord {
        rng.choose(&REINHOLD_WORDLIST).unwrap().clone()
    }
}
*/

impl fmt::Display for ReinholdWord {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let ReinholdWord(w) = self;
        write!(f, "{}", w)
    }
}

impl rand::distributions::Distribution<MiniLockWord> for rand::distributions::Standard {
    fn sample<R: rand::Rng + ?Sized>(&self, mut rng: &mut R) -> MiniLockWord {
        *MINILOCK_WORDLIST.choose(&mut rng).unwrap()
    }
}

/*
impl rand::Rand for MiniLockWord {
    fn rand<R: rand::Rng>(rng: &mut R) -> MiniLockWord {
        rng.choose(&MINILOCK_WORDLIST).unwrap().clone()
    }
}
*/

impl fmt::Display for MiniLockWord {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let MiniLockWord(w) = self;
        write!(f, "{}", w)
    }
}


diceware_wordlists/src/lib.rs
use std::str::FromStr;

include!(concat!(env!("OUT_DIR"), "/diceware.rs"));

#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
pub enum Wordlist {
    #[cfg(feature = "beale")]
    Beale,
    #[cfg(feature = "reinhold")]
    Reinhold,
    #[cfg(feature = "minilock")]
    Minilock,
    #[cfg(feature = "efflong")]
    EffLong,
    #[cfg(feature = "effshort1")]
    EffShort1,
    #[cfg(feature = "effshort2")]
    EffShort2,
}

impl Wordlist {
    pub fn get_list(&self) -> &'static [&'static str] {
        match self {
            #[cfg(feature = "beale")]
            Self::Beale => &BEALE_WORDLIST,
            #[cfg(feature = "reinhold")]
            Self::Reinhold => &REINHOLD_WORDLIST,
            #[cfg(feature = "minilock")]
            Self::Minilock => &MINILOCK_WORDLIST,
            #[cfg(feature = "efflong")]
            Self::EffLong => &EFF_LONG_WORDLIST,
            #[cfg(feature = "effshort1")]
            Self::EffShort1 => &EFF_SHORT_WORDLIST_1,
            #[cfg(feature = "effshort2")]
            Self::EffShort2 => &EFF_SHORT_WORDLIST_2_0,
        }
    }
}

impl FromStr for Wordlist {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            #[cfg(feature = "efflong")]
            "efflong" => Ok(Self::EffLong),
            #[cfg(feature = "reinhold")]
            "reinhold" => Ok(Self::Reinhold),
            #[cfg(feature = "beale")]
            "beale" => Ok(Self::Beale),
            #[cfg(feature = "minilock")]
            "minilock" => Ok(Self::Minilock),
            #[cfg(feature = "effshort1")]
            "effshort1" => Ok(Self::EffShort1),
            #[cfg(feature = "effshort2")]
            "effshort2" => Ok(Self::EffShort2),
            _ => Err("Unknown Wordlist"),
        }
    }
}

// XXX If the `efflong` feature is not chosen, which wordlist should be next in line?
// #[cfg(feature = "efflong")]
impl Default for Wordlist {
    fn default() -> Self {
        Self::EffLong
    }
}


diceware_wordlists/build.rs
use std::env;
use std::fs::File;
use std::io::Read;
use std::io::Write;
use std::path::Path;
use std::string::String;

fn make_diceware_wordlist(array_name: String) -> impl Fn(&str) -> String {
    move |contents: &str| {
        // 7776 words = 6*6*6*6*6; five 6 faced dice throws.
        format!("pub static {}: [&str; 7776] = [\n", array_name)
            + &contents
                .split('\n')
                .skip(2)
                .take(6 * 6 * 6 * 6 * 6)
                .map(|s| s.split_once('\t').unwrap().1)
                .map(|s| {
                    s.chars()
                        .map(|c| {
                            if c == '"' {
                                "\\\"".to_owned()
                            } else {
                                c.to_string()
                            }
                        })
                        .collect::<String>()
                })
                .map(|s| format!("    \"{}\",\n", s))
                .collect::<String>()
            + "];\n\n"
    }
}

fn make_minilock_wordlist(contents: &str) -> std::string::String {
    // 58110 words in the MiniLock wordlist.
    "pub static MINILOCK_WORDLIST: [&str; 58110] = [\n".to_owned()
        + &contents[1023..543718]
            .split(',')
            .map(|x| format!("    \"{}\",\n", x))
            .collect::<String>()
        + "];\n"
}

fn make_eff_long_wordlist(contents: &str) -> String {
    // 7776 words = 6*6*6*6*6; five 6 faced dice throws.
    "pub static EFF_LONG_WORDLIST: [&str; 7776] = [\n".to_owned()
        + &contents
            .split('\n')
            .take(6 * 6 * 6 * 6 * 6)
            .map(|x| x.split('\t').nth(1).unwrap())
            .map(|x| format!("    \"{}\",\n", x))
            .collect::<String>()
        + "];\n"
}

fn make_eff_short_wordlist(array_name: String) -> impl Fn(&str) -> String {
    move |contents: &str| {
        // 1296 words = 6*6*6*6; five 6 faced dice throws.
        format!("pub static {}: [&str; 1296] = [\n", array_name)
            + &contents
                .split('\n')
                .take(6 * 6 * 6 * 6)
                .map(|x| x.split('\t').nth(1).unwrap())
                .map(|x| format!("    \"{}\",\n", x))
                .collect::<String>()
            + "];\n"
    }
}

fn build_wordlist(
    wordlist_file_path: &str,
    destination_file: &mut File,
    make_wordlist: &dyn Fn(&str) -> String,
) {
    let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    println!("{}", manifest_dir);
    let wordlist_file_path = Path::new(&manifest_dir).join(wordlist_file_path);

    let mut wordlist_file = File::open(&wordlist_file_path).unwrap();

    let mut wordlist_string = String::new();

    wordlist_file.read_to_string(&mut wordlist_string).unwrap();

    destination_file
        .write_all(make_wordlist(&wordlist_string).as_bytes())
        .unwrap();
}

fn main() {
    let out_dir = env::var("OUT_DIR").unwrap();
    let dest_path = Path::new(&out_dir).join("diceware.rs");
    let mut destination_file = File::create(&dest_path).unwrap();

    #[cfg(feature = "beale")]
    build_wordlist(
        "wordlists/beale.wordlist.asc",
        &mut destination_file,
        &make_diceware_wordlist("BEALE_WORDLIST".to_string()),
    );
    #[cfg(feature = "reinhold")]
    build_wordlist(
        "wordlists/diceware.wordlist.asc",
        &mut destination_file,
        &make_diceware_wordlist("REINHOLD_WORDLIST".to_string()),
    );
    #[cfg(feature = "minilock")]
    build_wordlist(
        "wordlists/phrase.js",
        &mut destination_file,
        &make_minilock_wordlist,
    );
    #[cfg(feature = "efflong")]
    build_wordlist(
        // They call it "EFF's Long Wordlist" on https://www.eff.org/dice, not "EFF's Large
        // Wordlist", although the file name is different, so I will call it "long" rather than
        // "large".
        "wordlists/www.eff.org/files/2016/07/18/eff_large_wordlist.txt",
        &mut destination_file,
        &make_eff_long_wordlist,
    );
    #[cfg(feature = "effshort1")]
    build_wordlist(
        "wordlists/www.eff.org/files/2016/09/08/eff_short_wordlist_1.txt",
        &mut destination_file,
        &make_eff_short_wordlist("EFF_SHORT_WORDLIST_1".to_string()),
    );
    #[cfg(feature = "effshort2")]
    build_wordlist(
        "wordlists/www.eff.org/files/2016/09/08/eff_short_wordlist_2_0.txt",
        &mut destination_file,
        &make_eff_short_wordlist("EFF_SHORT_WORDLIST_2_0".to_string()),
    );
}
